### PR TITLE
feat: Added toString method to all the drags events message handlers

### LINKED ISSUE #2007: Add `toString()` methods to Event objects
# What could be improved

The `Event` objects in `flame/lib/src/events/messages/` are pure data objects, and can benefit from having meaningful `toString()` representations.

# Why should this be improved

Occasionally users need to debug events, and in that case printing the event object into a console is the easiest. However, without a `toString()` method, the objects are printed as "Instance of <DragStartEvent>", which is nearly useless. A good `toString` method would also show the event coordinates, pointer id, etc.

# Any risks?

None, since `toString()` is used mostly for debugging.

### PR DESCRIPTION
**Pull request by sumit7577**
# Description
Added toString method to all Drag Event message Handlers in flame/lib/src/events/messages/  which were returning pure data objects and were not much meaningful

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have followed the [Contributor Guide] when preparing my PR.
- [ ] I have updated/added tests for ALL new/updated/fixed functionality.
- [ ] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

- [ ] Yes, this PR is a breaking change.
- [x] No, this PR is not a breaking change.

## Related Issues
Closes #2007

[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

File: tutorials/space_shooter/test/widget_test.dart [hunk 1]
```diff
@@ -0,0 +1,30 @@
+// This is a basic Flutter widget test.
```
Comment (spydon) [inline tutorials/space_shooter/test/widget_test.dart:1]: This file should be removed

File: tutorials/space_shooter/lib/steps/1_getting_started/1_flame_game/code.dart [hunk 2]
```diff
@@ -4,7 +4,6 @@ import 'package:flutter/material.dart';
 
 class Player extends PositionComponent {
   static final _paint = Paint()..color = Colors.white;
-
```
Comment (spydon) [inline tutorials/space_shooter/lib/steps/1_getting_started/1_flame_game/code.dart:7]: Unnecessary change and unrelated to this PR.

File: packages/flame/lib/src/events/messages/drag_update_event.dart [hunk 3]
```diff
@@ -36,4 +36,9 @@ class DragUpdateEvent extends PositionEvent {
       ),
     );
   }
+
+  @override
+  String toString() =>
+      'DragUpdateEvent(pointerId: $pointerId,timestamp: $timestamp'
```
Comment (spydon) [inline packages/flame/lib/src/events/messages/drag_update_event.dart:42]: Make sure that all commas have a space after them

Comment (spydon) [reply packages/flame/lib/src/events/messages/drag_update_event.dart:42]: timestamp also needs a comma after it since delta continues on the next line

File: packages/flame/lib/src/events/messages/drag_update_event.dart [hunk 4]
```diff
@@ -36,4 +36,9 @@ class DragUpdateEvent extends PositionEvent {
...
+      'delta: $delta, _nanPoint: $_nanPoint, devicePosition:$devicePosition, canvasPosition:$canvasPosition)';
```
Comment (spydon) [inline packages/flame/lib/src/events/messages/drag_update_event.dart:43]: Too long line, remember that the max line length is 80.

File: tutorials/space_shooter/lib/steps/1_getting_started/1_flame_game/code.dart (hunk 2 above)
Comment (sumit7577) [reply tutorials/space_shooter/lib/steps/1_getting_started/1_flame_game/code.dart:7]: This got changed while testing i fixed it now

Comment (Unknownjohnny) [reply tutorials/space_shooter/lib/steps/1_getting_started/1_flame_game/code.dart:7]: Ok sir

File: packages/flame/lib/src/events/messages/drag_cancel_event.dart [hunk 5]
```diff
@@ -11,4 +11,6 @@ class DragCancelEvent extends Event {
   final int pointerId;
 
   DragEndEvent toDragEnd() => DragEndEvent(pointerId, DragEndDetails());
+  @override
+  String toString() => 'DragCancelEvent(pointerId:$pointerId)';
```
Comment (st-pasha) [inline packages/flame/lib/src/events/messages/drag_cancel_event.dart:15]: Nit: add space after `:`

File: packages/flame/lib/src/events/messages/drag_end_event.dart [hunk 6]
```diff
@@ -20,4 +20,7 @@ class DragEndEvent extends Event {
       ),
     );
   }
+
+  @override
+  String toString() => 'DragEndEvent(pointerId:$pointerId, Velocity:$velocity)';
```
Comment (st-pasha) [inline packages/flame/lib/src/events/messages/drag_end_event.dart:25]: `Velocity` should be lowercase, also spaces after `:`s

File: packages/flame/lib/src/events/messages/drag_start_event.dart [hunk 7]
```diff
@@ -34,4 +34,9 @@ class DragStartEvent extends PositionEvent {
       ),
     );
   }
+
+  @override
+  String toString() =>
+      'DragStartEvent(pointerId: $pointerId, deviceKind: $deviceKind, '
+      'canvasPosition: $canvasPosition, devicePosition: $devicePosition)';
```
Comment (st-pasha) [inline packages/flame/lib/src/events/messages/drag_start_event.dart:41]: The `devicePosition` is rarely useful, since the games almost always run in fullscreen mode. Instead, let's replace it with the `localPosition`. Also, `canvasPosition`/`localPosition` are more important than `pointerId`/`deviceKind`, so let's switch them around.

File: packages/flame/lib/src/events/messages/drag_update_event.dart [hunk 8]
```diff
@@ -36,4 +36,10 @@ class DragUpdateEvent extends PositionEvent {
       ),
     );
   }
+
+  @override
+  String toString() =>
+      'DragUpdateEvent(pointerId: $pointerId, timestamp: $timestamp, '
+      'delta: $delta, _nanPoint: $_nanPoint, devicePosition:$devicePosition, '
+      'canvasPosition:$canvasPosition)';
```
Comment (st-pasha) [inline packages/flame/lib/src/events/messages/drag_update_event.dart:44]: `_nanPoint` is useless to the user (it's a constant). 
Similarly to DragStartEvent, let's have `canvasPosition+localPosition` first, followed by `delta`, followed by `pointerId`/`timestamp`.

File: packages/flame/lib/src/events/messages/tap_cancel_event.dart [hunk 9]
```diff
@@ -19,4 +19,7 @@ class TapCancelEvent extends Event {
   /// The id of the event that has been cancelled. This id corresponds to the
   /// id of the previous [TapDownEvent].
   final int pointerId;
+
+  @override
+  String toString() => 'TapCancelEvent(pointerId;$pointerId)';
```
Comment (st-pasha) [inline packages/flame/lib/src/events/messages/tap_cancel_event.dart:24]: typo: `;`

File: packages/flame/lib/src/events/messages/tap_down_event.dart [hunk 10]
```diff
@@ -44,4 +44,9 @@ class TapDownEvent extends PositionEvent {
       ),
     );
   }
+
+  @override
+  String toString() =>
+      'TapDownEvent(pointerId;$pointerId, deviceKind:$deviceKind, '
+      'canvasPosition:$canvasPosition, devicePosition:$devicePosition)';
```
Comment (st-pasha) [inline packages/flame/lib/src/events/messages/tap_down_event.dart:51]: See comment for DragStartEvent

File: packages/flame/lib/src/events/messages/tap_up_event.dart [hunk 11]
```diff
@@ -36,4 +36,9 @@ class TapUpEvent extends PositionEvent {
       ),
     );
   }
+
+  @override
+  String toString() =>
+      'TapUpEvent(pointerId;$pointerId, deviceKind:$deviceKind, '
+      'canvasPosition:$canvasPosition, devicePosition:$devicePosition)';
```
Comment (st-pasha) [inline packages/flame/lib/src/events/messages/tap_up_event.dart:43]: See comment for DragStartEvent

**Review by st-pasha — COMMENTED** [review]
It might also be useful to add some tests that the string representation works as expected. These can be added directly into `has_draggable_components_test.dart`. Say, the very first test when receives a drag start event, it just increments a counter. But you could make it store the event `e` into a local variable, and then check that it stringifies properly.

File: packages/flame/lib/src/events/messages/drag_cancel_event.dart (hunk 5 above)
Comment (ufrshubham) [inline packages/flame/lib/src/events/messages/drag_cancel_event.dart:15]:
```suggestion
  String toString() => '$DragCancelEvent(pointerId:$pointerId)';
```
Will help in keeping class name in this string in-sync with the actual class name in future as well.

File: packages/flame/lib/src/events/messages/drag_end_event.dart (hunk 6 above)
Comment (ufrshubham) [inline packages/flame/lib/src/events/messages/drag_end_event.dart:25]:
```suggestion
  String toString() => '$DragEndEvent(pointerId:$pointerId, Velocity:$velocity)';
```
Will help in keeping class name in this string in-sync with the actual class name in future as well.

File: packages/flame/lib/src/events/messages/drag_start_event.dart [hunk 12]
```diff
@@ -34,4 +34,9 @@ class DragStartEvent extends PositionEvent {
       ),
     );
   }
+
+  @override
+  String toString() =>
+      'DragStartEvent(pointerId: $pointerId, deviceKind: $deviceKind, '
```
Comment (ufrshubham) [inline packages/flame/lib/src/events/messages/drag_start_event.dart:40]:
```suggestion
      '$DragStartEvent(pointerId: $pointerId, deviceKind: $deviceKind, '
```
Will help in keeping class name in this string in-sync with the actual class name in future as well.

File: packages/flame/lib/src/events/messages/drag_update_event.dart [hunk 13]
```diff
@@ -36,4 +36,10 @@ class DragUpdateEvent extends PositionEvent {
       ),
     );
   }
+
+  @override
+  String toString() =>
+      'DragUpdateEvent(pointerId: $pointerId, timestamp: $timestamp, '
```
Comment (ufrshubham) [inline packages/flame/lib/src/events/messages/drag_update_event.dart:42]:
```suggestion
      '$DragUpdateEvent(pointerId: $pointerId, timestamp: $timestamp, '
```
Will help in keeping class name in this string in-sync with the actual class name in future as well.

File: packages/flame/lib/src/events/messages/tap_cancel_event.dart (hunk 9 above)
Comment (ufrshubham) [inline packages/flame/lib/src/events/messages/tap_cancel_event.dart:24]:
```suggestion
  String toString() => '$TapCancelEvent(pointerId;$pointerId)';
```
Will help in keeping class name in this string in-sync with the actual class name in future as well.

File: packages/flame/lib/src/events/messages/tap_down_event.dart [hunk 14]
```diff
@@ -44,4 +44,9 @@ class TapDownEvent extends PositionEvent {
       ),
     );
   }
+
+  @override
+  String toString() =>
+      'TapDownEvent(pointerId;$pointerId, deviceKind:$deviceKind, '
```
Comment (ufrshubham) [inline packages/flame/lib/src/events/messages/tap_down_event.dart:50]:
```suggestion
      '$TapDownEvent(pointerId;$pointerId, deviceKind:$deviceKind, '
```
Will help in keeping class name in this string in-sync with the actual class name in future as well.

File: packages/flame/lib/src/events/messages/tap_up_event.dart [hunk 15]
```diff
@@ -36,4 +36,9 @@ class TapUpEvent extends PositionEvent {
       ),
     );
   }
+
+  @override
+  String toString() =>
+      'TapUpEvent(pointerId;$pointerId, deviceKind:$deviceKind, '
```
Comment (ufrshubham) [inline packages/flame/lib/src/events/messages/tap_up_event.dart:42]:
```suggestion
      '$TapUpEvent(pointerId;$pointerId, deviceKind:$deviceKind, '
```
Will help in keeping class name in this string in-sync with the actual class name in future as well.

**Review by ufrshubham — COMMENTED** [review]
A minor change to makes sure class names and the name in toString does not deviate in future.

File: packages/flame/lib/src/events/messages/drag_cancel_event.dart (hunk 5 above)
Comment (st-pasha) [reply packages/flame/lib/src/events/messages/drag_cancel_event.dart:15]: Not sure it's worth it, we are not planning to rename these events in the foreseeable future.

Comment (ufrshubham) [reply packages/flame/lib/src/events/messages/drag_cancel_event.dart:15]: Agreed. It was just suggestion anyways. 

But it might help in avoiding the classic copy-paste problem too. If a new event class is added and someone just implements `toString` by copying it from an existing event class, they would actually get an import error.

Not too big of a use case though 😅

Comment (st-pasha) [reply packages/flame/lib/src/events/messages/drag_cancel_event.dart:15]: Yeah, it just looks a bit weird -- never seen anyone doing it like that :)

I'm gonna close the suggestions.

**Review by erickzanardo — APPROVED** [review]
LGTM once all nits are resolved.

**Comment by spydon** [comment]
@sumit7577 any update on this PR?

**Comment by sumit7577** [comment]
> @sumit7577 any update on this PR?

File: packages/flame/lib/src/events/messages/drag_cancel_event.dart [hunk 16]
```diff
@@ -11,4 +11,6 @@ class DragCancelEvent extends Event {
   final int pointerId;
 
   DragEndEvent toDragEnd() => DragEndEvent(pointerId, DragEndDetails());
+  @override
```
[the remark this answers is not in the document]
Comment (spydon) [reply packages/flame/lib/src/events/messages/drag_cancel_event.dart:14]: @sumit7577 this is not resolved yet.

**Review by spydon — APPROVED** [review]

**Comment by spydon** [comment]
Good job!
